docs/static-evaluation.md

```diff
@@ -0,0 +1,446 @@
+
+When evaluating an Attribute/Block into a value, any missing reference must be properly reported in a way that the user can easily debug and understand. For example, a user may try to use a `local` that depends on a resource's value in a module's source. The user must then be told that the `local` can not be used in the module source field as it depends on a resource which is not available during the config loading process. Variables used through the module tree must also be passed with their associated information, such as their references.
+
+TODO add a note that the current evaluator does not support dependency chains. Why do we need these? Forward reference?
+TODO example for this, as well as an example error message.
+TODO the whole previous paragraph should be rewritten to describe user experience, not technical details.
+
+Implementing this initial concept will allow many of the [solutions](#solutions) below to be fully implemented.  However, there are some limitations due to module expansion which are worth considering.
+
+## Additional Complexity due to Module Expansion
+
+The concepts of `for_each` and `count` were grafted on to the codebase in a way that has added significant complexity and limitations. When a module block contains a `for_each` or `count` all of the nodes (resources/variables/locals/etc...) will be created multiple times, one copy per "instance".
+
+One common example of a limitation would be to use different providers for different module instances, which does not work today:
+
+```hcl
+# main.tf
+module "mod" {
+        for_each = {"us" = "first", "eu" = "second"}
+        source = "./mod"
+        name = each.key
+        providers {
+          aws = provider.aws[each.value] # Currently only hard-coded values are supported here
+        }
+}
+```
+
+As the provider requirements are baked into the module itself, the multiple "instances" don't have any concept of providers per instance. This becomes even more complex when you consider that these providers might be passed through a complex tree of modules before they are directly used.
+
+There are two potential solutions detailed below in [Provider Iteration](#provider-iteration), each with their own tradeoffs.
+
+A related problem is using `each.key/value` or `count` in module sources.  The sources don't have any concept of "module instances" from `for_each/count` and have similar limitations to provider iteration.  This a much lower priority item that is likely not going to be considered for the foreseeable future. TODO reference module expansion document.
+
+## Plan of Attack
+
+Between the initial implementation, the simple Solutions, and the provider complexity, we are talking about a significant amount of work likely spread across multiple releases.
```

Comment:
If this feature will have multiple releases, can a structured plan be created to detail which features will be rolled out in each release?

Reply:
That's a good question.  I think I need to re-structure this RFC based on the conversations that have been happening around the new RFC process PR.  I think it is actually going to be two different RFCs in the end to properly scope the work.